analysis/src/ProcessCmt.ml

```diff
@@ -793,3 +793,77 @@ let fileForModule moduleName ~package =
   | None ->
     Log.log ("No path for module " ^ moduleName);
     None
+
+(* Collect top-level exception constructors from typedtree/CMT file. *)
+let exceptionsForCmt ~cmt : (string * bool) list =
+  match Shared.tryReadCmt cmt with
+  | None -> []
+  | Some infos ->
+    let by_name : (string, bool) Hashtbl.t = Hashtbl.create 16 in
+    let add_ext (ext : Typedtree.extension_constructor) : unit =
+      let name = ext.ext_name.txt in
+      let hasArgs =
+        match ext.ext_kind with
+        | Text_decl (Cstr_tuple args, _ret) -> args <> []
+        | Text_decl (Cstr_record fields, _ret) -> fields <> []
+        | Text_rebind _ -> true
+      in
+      let prev =
+        match Hashtbl.find_opt by_name name with
+        | Some b -> b
+        | None -> false
+      in
+      Hashtbl.replace by_name name (prev || hasArgs)
+    in
+    (* Only collect top-level exception declarations (Tstr_exception/Tsig_exception).
+     Avoid picking up exceptions from Texp_letexception by tracking context. *)
```

Comment:
In theory I guess the most robust approach would be to collect and use whatever defined exceptions are in scope at the current position. Collecting at the top level is good and simple, but you could also define them in sub modules, and so on.

It might be worth exploring making exceptions part of the tracked scope instead, and track them just like we track types and values today when doing completion. Then we could just extend the current mechanism a bit, and hopefully get the correct scope tracking for free.